### DISCUSSION TITLE
Migration (from 0.27 to 0.41)

### DISCUSSION THREAD

**Discussion by schmitzli** [nushell/nushell#4213] [Q&A] [answered]
Hi,

i really like Nu shell and want to migrate to the latest version.
i have about 15 k of definitions in a startup file (version 0.27).
The migration is quite frustrating as i don't have any tests and everything seems to be broken...

To contribute i'll first list the required changes that i performed to the script.

1. evaluation: **$(..)** is now just **(..)**
2. mathmode: **{= -1}** is now just **{-1}**
3. renamed: **count** translates to **length** 
4. **empty?** requires **-b** before an insert block
5. number to string: **str from** is removed, in my case i could use **format "{size}"**
6. 7zip:  **7z l $zip** can still be used with shell escape **^7z l $zip**
7. headerless:  **from tsv --headerless** is now **from tsv --noheaders**
8. output redirect:  **mutool info $it.file > muinfo.txt** breaks because of the > : **save -raw** still has a buffer bug, so i think i can do batch call for this

With this i covered now about 3/4 of the definitions, but

Somehow iterator use and variable scope has old and new issues (or i don't use it correctly).

A: $it not recognized after adding -b before the if block
this one is not working anymore and fails to recognize $it.size
`def add-ext [] { each { insert ext {echo $it.path | parse -r .*\.(?P<ex>(\w){1,5})$ | get ex | str downcase}} | empty? ext {if $it.size == -1 {echo dir} {echo ext}} }`

without the **if** it's working, i now do `ext -b { echo $it.type | str downcase } }` as a workaround

B: **save** not working
here i'm really stuck
` def add-sum [] {open info.toml | insert b3sum { let f = $it.file ; calc-hash $f | str collect } | save}`

`Welcome to Nushell 0.41.0 (type 'help' for more info)
> ws
> source .\startup41.nu
> cd .\test\
> add-sum
error: Save requires a filepath
  ┌─ shell:5:1
  │
5 │
  │   needs path

error: Variable not in scope
  ┌─ shell:5:1
  │
5 │
  │   unknown variable: $it

> open .\info.toml | each { calc-hash $it.file }
2cb949be2b7e86ecbe1fa862be72bcef60d8a484fc47ed5f7e317ba729b08442
>
`
i reduced the inner block to just **calc-hash $it.file** but without improvement.

If you know more conversion steps, please add them.
If you have any hint to solve issues A and B please tell.

Thank you very much

**Comment by kubouch** [comment, marked as answer]
Hi, that is a lot of code you're porting. You're right, we made a big syntax update I think around version 0.32. However, updating now for 0.41 might not be the best idea since we're in the middle of another update that will replace Nushell's current parser and engine and that will change basically everything. You can check it here: https://github.com/nushell/engine-q. While we aim for most commands working the same as before, there will be breaking changes and some underlying systems will work in a fundamentally different way. I'd suggest either waiting until it gets merged, or trying out the "engine-q" repository which still has some missing commands, however.

The scope issues are now fixed in the new engine and also the error reporting is now accurate so that will hopefully make debugging easier.

If you came across some specific issues (like the `$it` missing in the `insert` block), I'd suggest you try them with the engine-q and report an issue if it does not work there.

For the `save` in B, I think you need to give it a file name.

**Reply by hustcer** [reply]
@kubouch when will engine-q be merged into nushell ? Is there a rough plan thanks

**Reply by schmitzli** [reply]
Thank you, it's probably the best to wait for the engine-q release with any migration plans.
There is a list of breaking changes here [engine-q breaking chnages](https://github.com/nushell/engine-q/issues/522)
For anyone doing some script migration, it's probably a good idea to think about some integration tests.
In my case i'll prep some data, run the commands, save to file and compare with diff the test output to an original.

**Reply by kubouch** [reply]
@hustcer We aim for early next year. We're getting pretty close actually. Most commands are ported, but we need to spend some time polishing things.

@schmitzli The breaking changes list is incomplete and WIP. We just started putting it together so I expect a lot of things are missing. About the integration tests, you suggest Nushell should have more integration tests or that you need to make some for yourself?

**Reply by schmitzli** [reply]
i meant integration tests for my scripts, as several changes might be necessary when moving to a new version
